.github/workflows/test-tpg.yml

```diff
@@ -14,36 +20,90 @@ on:
         required: false
         default: 'terraform-provider-google'
       branch:
-        description: 'The branch or sha to execute against'
+        description: 'The branch or sha of the provider execute against'
+        required: true
+      sha:
+        description: "The commit SHA in magic-modules repository where the status result will be posted"
         required: true
 
+concurrency:
+  group: test-tpg-${{ github.event.inputs.owner }}-${{ github.event.inputs.repo }}-${{ github.event.inputs.branch }}
+  cancel-in-progress: true
+
 jobs:
-  build-and-test:
+  build-and-unit-test:
```

Comment:
I think this should be split into separate jobs so that we can more accurately capture the [dependencies](https://docs.github.com/en/actions/using-workflows/about-workflows#creating-dependent-jobs) between them and have better handling of partial failures than can be captured with steps only.

For example, lint checks, doc tests, and unit tests should all be run even if one of them fails, and they can run in parallel.

I suspect that you might be able to fetch all the job html_urls in the setup phase and then either update all the statuses at the end or at a per-job level - either is fine as long as the updating of status happens always.

So something like:

```
jobs:
  setup:
    # get jobs list & post pending statuses & checkout & look for code changes & other setup
  build:
    needs: setup
  lint:
    if: ${{ !failure() && needs.setup.outputs.has_changes == 'true' }}
    needs: [setup, build]  # Is this true or does it just need setup?
  test:
    if: ${{ !failure() && needs.setup.outputs.has_changes == 'true' }}
    needs: [setup, build]  # Is this true or does it just need setup? 
  update_statuses:
    if: ${{ always() }}
    needs: [lint, test]
    steps:  # update github statuses for the lint & test jobs
```

What do you think?

Reply:
I disagree. Jobs don't share context so you would need to re clone the repository for every job. The setup step you have here won't provide the repository for all of the other jobs.

Lint, and Docs already run even if the build fails. The only thing that doesn't run is the unit tests.

The only way to share data between jobs is through artifacts or cache which would incur latency
https://levelup.gitconnected.com/github-actions-how-to-share-data-between-jobs-fc1547defc3e

Reply:
> Jobs don't share context so you would need to re clone the repository for every job

ah, yeah, that makes sense.

> The only way to share data between jobs...

That's true in terms of binaries, artifacts, etc. But you can pass simple values using the `needs` context. https://docs.github.com/en/actions/learn-github-actions/contexts#needs-context

... but that's not really relevant here since, as you pointed out, this needs to all be inside one job to share the checkout.

Reply:
oo this is pretty useful for a multi job design ! I'll keep it in mind ! :)

Reply:
resolving for now